### PR TITLE
Add hub for realtime multiplayer

### PR DESCRIPTION
**Pull request by peppy**
Initial logic pass. Doesn't yet have an actual client implementation (so serialization has not been tested sufficiently yet).

Potentially will be split out into its own project eventually, but made too much sense to keep it here while we can.

- [x] Depends on https://github.com/ppy/osu/pull/11127 (use local game reference to test)

### CODE REVIEW

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (bdach) [inline osu.Server.Spectator/Hubs/MultiplayerHub.cs:312]: Shouldn't this be checking `FinishedPlay`, or just blocking this on the client side entirely? I guess it'll be more apparent after a client implementation lands...

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (peppy) [reply osu.Server.Spectator/Hubs/MultiplayerHub.cs:19]: i switched away from ConcurrentDictionary for that reason specifically. Doesn't seem like the class we want here.

Keeping locking simple until performance is actually a problem (doubt it will ever be).

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (peppy) [reply osu.Server.Spectator/Hubs/MultiplayerHub.cs:312]: good catch, should be denied. i changed how the flow worked halfway through.

File: osu.Server.Spectator.Tests/MultiplayerFlowTests.cs
Comment (peppy) [reply osu.Server.Spectator.Tests/MultiplayerFlowTests.cs:373]: doesn't work well as `await` calls can not occur inside a delegate.

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (bdach) [inline osu.Server.Spectator/Hubs/MultiplayerHub.cs:262]: From what I gathered, the user can always back out to `Idle`, in light of which this should probably be checking

```suggestion
                    if (room.Users.All(u => u.State == MultiplayerUserState.Loaded))
```

and have a special case for all users re-entering `Idle`. Here's a failing test case:

```csharp
        [Fact]
        public async Task AllUsersBackingOutFromLoadCancelsTransitionToPlay()
        {
            var room = await hub.JoinRoom(room_id);

            setUserContext(mockContextUser2);
            await hub.JoinRoom(room_id);

            await hub.ChangeState(MultiplayerUserState.Ready);

            setUserContext(mockContextUser1);
            await hub.ChangeState(MultiplayerUserState.Ready);
            await hub.StartMatch();

            Assert.Equal(MultiplayerRoomState.WaitingForLoad, room.State);

            await hub.ChangeState(MultiplayerUserState.Idle);
            setUserContext(mockContextUser2);
            await hub.ChangeState(MultiplayerUserState.Idle);

            Assert.Equal(MultiplayerRoomState.Open, room.State);
        }
```

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (bdach) [inline osu.Server.Spectator/Hubs/MultiplayerHub.cs:265]: Upon reading `StartMatch()`, it is possible that only part of the room has previously entered gameplay, which in light of this code creates the possibility of wrongly changing state of users that have never readied up. `Room` should probably have a `UsersInGameplay` or similar (`ReadiedUpUsers`?)

Failing test case for this:

```csharp
        [Fact]
        public async Task OnlyReadiedUpUsersTransitionToPlay()
        {
            var room = await hub.JoinRoom(room_id);
            await hub.ChangeState(MultiplayerUserState.Ready);

            setUserContext(mockContextUser2);
            await hub.JoinRoom(room_id);

            setUserContext(mockContextUser1);
            await hub.StartMatch();
            Assert.Equal(MultiplayerRoomState.WaitingForLoad, room.State);
            Assert.Single(room.Users, u => u.State == MultiplayerUserState.WaitingForLoad);
            Assert.Single(room.Users, u => u.State == MultiplayerUserState.Idle);

            await hub.ChangeState(MultiplayerUserState.Loaded);
            Assert.Single(room.Users, u => u.State == MultiplayerUserState.Playing);
            Assert.Single(room.Users, u => u.State == MultiplayerUserState.Idle);
        }
```

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (bdach) [inline osu.Server.Spectator/Hubs/MultiplayerHub.cs:280]: See above pair of comments, but with `FinishedPlay`.

**Review by bdach — COMMENTED** [review]
Sorry if these are getting annoying because this isn't finished yet - hard for me to tell what state this is in, just wanting to help out, is all.

File: osu.Server.Spectator/Hubs/MultiplayerHub.cs
Comment (peppy) [reply osu.Server.Spectator/Hubs/MultiplayerHub.cs:265]: This is really weird, I thought I already had this implemented correctly..

**Comment by peppy** [comment]
Awesome feedback @bdach, thanks!

**Review by bdach — APPROVED** [review]
A little tired as I read this, but looks like a solid initial pass at this stage to me :+1:

**Review by smoogipoo — APPROVED** [review]
Lgtm

**Pull request merged**
